internal/rpc/core/mempool.go

```diff
@@ -74,11 +74,17 @@ func (env *Environment) BroadcastTxCommit(ctx context.Context, tx types.Tx) (*co
 
 	r := (<-resCh).GetCheckTx()
 	if r.Code != abci.CodeTypeOK {
+		if r.MempoolError == "" {
+			err = errors.New("transaction did not succeed")
+		} else {
+			err = errors.New(r.MempoolError)
+		}
```

Comment:
Possible alternative:
```suggestion
		err = fmt.Errorf("transaction is not valid (%s)", r.MempoolError)
```

Rationale: "Success" kind of implies execution, but this is just validation. And in case of a mempool error, I could see having it unqualified by context be confusing in logs. YMMV, though.